applications/parameter_sweep.md

```diff
@@ -0,0 +1,229 @@
+
+First, we import the following libraries:
+```python
+from magicgui import magicgui, register_type
+from magicgui._qt import QDoubleSlider
+from napari import Viewer, gui_qt, layers
+from skimage import data
+from skimage.filters import threshold_sauvola, threshold_local, threshold_niblack
+from math import floor
+```
+
+Next, we define the methods `get_layers` and `show_layer_result` to obtain the layers in the napari viewer, as well as add any new layers to display the result of thresholding:
+```python
+def get_layers(gui, layer_type):
+    try:
+        viewer = gui.parent().qt_viewer.viewer
+        return tuple(l for l in viewer.layers if isinstance(l, layer_type))
+    except AttributeError:
+        return ()
+
+def show_layer_result(gui, result, return_type) -> None:
+    if result is None:
+        return
+    try:
+        viewer = gui.parent().qt_viewer.viewer
+    except AttributeError:
+        return
+    try:
+        viewer.layers[gui.result_name].data = result
+    except KeyError:
+        adder = getattr(viewer, f"add_{return_type.__name__.lower()}")
+        adder(data=result, name=gui.result_name)
+
+register_type(layers.Layer, choices=get_layers, return_callback=show_layer_result)
+
+```
+
+We then create a viewer with napari, and add the `data.page()` image layer to try out parameter sweep with different thresholding techniques. We use the `magicgui` decorator for the method `apply_threshold` to turn it into a magicgui. 
+
+`auto_call = True` indicates magicgui to call `apply_threshold` whenever the value of a parameter changes. The QDoubleSlider is used to pass the size parameter to the function, and we offer the choices to threshold an image layer using the adaptive, Sauvola and Niblack techniques:
+
+```python
+with gui_qt():
+    # creating a viewer and adding the data.page() image layer
+    viewer = Viewer()
+    viewer.add_image(data.page(), name="page")
+    
+    # passing size as (2* floor(size/2)+1), as block-size and window size have to be odd natural numbers
+    @magicgui(
+        auto_call=True,
+        size={"widget_type": QDoubleSlider, "maximum": 99, "fixedWidth": 400},
+        technique={"choices": ["adaptive", "niblack", "sauvola"]},
+    )
+    def apply_threshold(layer: layers.Image, size=1, technique="adaptive") -> None:
+        if layer:
+            if technique=="adaptive":
+                return layer.data > threshold_local(layer.data, (2* floor(size/2)+1))
+            elif technique=="sauvola":
+                return layer.data > threshold_sauvola(layer.data, (2* floor(size/2)+1))
+            elif technique=="niblack":
+                return layer.data > threshold_niblack(layer.data, (2* floor(size/2)+1))
```

Comment:
the entire example (including the `register_type` bit above and the `show_result` bit below) can be reduced to this:

```python
from magicgui import magicgui
from magicgui._qt import QDoubleSlider
from napari import Viewer, gui_qt, layers
from skimage import data, filters
from math import floor

with gui_qt():
    # creating a viewer and adding the data.page() image layer
    viewer = Viewer()
    viewer.add_image(data.page(), name="page")

    # passing size as (2* floor(size/2)+1), as block-size and window size
    # have to be odd natural numbers
    @magicgui(
        auto_call=True,
        size={"widget_type": QDoubleSlider, "maximum": 99, "fixedWidth": 400},
        technique={"choices": ["local", "niblack", "sauvola"]},
    )
    def apply_threshold(layer: layers.Image, size=50, technique="local") -> layers.Image:
        if layer:
            func = getattr(filters, "threshold_" + technique)
            return layer.data > func(layer.data, (2 * floor(size / 2) + 1))

    gui = apply_threshold.Gui()
    gui.parentChanged.connect(gui.refresh_choices)
    viewer.layers.events.changed.connect(lambda x: gui.refresh_choices("layer"))
    viewer.window.add_dock_widget(gui)
```

Things to point out here:

- we don't need to register anything for the `napari.layers.Layer` type on the current master branch of napari (or any future versions), so let's keep that out of this tutorial
- I made the default `size=50` so that the initial image isn't black
- when you have a lot of functions with similar names and identical APIs like these threshold functions, you can reduce a lot of code using the `func = getattr(filters, ...)` line here... this also makes it very easy to add a new threshold type simply by adding a new item to the `choices` list.
- note that I added a return annotation of ` -> layers.Image` to the `apply_threshold` function ... that tells `magicgui` to add an Image layer to the viewer... so we can actually get rid of the `show_result` callback below.